Add documentation for creating a Source Selection Algorithm (#3215)
magento/devdocs: guides/v2.3/inventory/source-selection-algorithms.md

---
group: inventory
title: Source selection algorithms
---

> The **Source Selection Algorithm (SSA)** receives the total number of shippable SKUs and their quantities in an order as well as the order's shipping address. It recommends how to fulfill partial and full shipments. The merchant decides which business needs take precedence when deciding which shipping method to use:

Reviewer: >receives the total number of shippable SKUs and their quantities in an order as well as the order's shipping address. 

This statement is not really correct, in our current implementation, SSA accepts `InventoryRequestInterface` object.

```
namespace Magento\InventorySourceSelectionApi\Api;

/**
 * Returns source selection algorithm result for given Inventory Request
 *
 * @api
 */
interface SourceSelectionServiceInterface
{
    /**
     * @param \Magento\InventorySourceSelectionApi\Api\Data\InventoryRequestInterface $inventoryRequest
     * @param string $algorithmCode
     * @return \Magento\InventorySourceSelectionApi\Api\Data\SourceSelectionResultInterface
     */
    public function execute(
        \Magento\InventorySourceSelectionApi\Api\Data\InventoryRequestInterface $inventoryRequest,
        string $algorithmCode
    ): \Magento\InventorySourceSelectionApi\Api\Data\SourceSelectionResultInterface;
}

```

Which in its turn contains `Stock Id`  for which we want to apply calculation and list of requested items.

```
interface ItemRequestInterface extends \Magento\Framework\Api\ExtensibleDataInterface
{
    /**
     * Requested SKU
     *
     * @return string
     */
    public function getSku(): string;

    /**
     * Requested Product Quantity
     *
     * @return float
     */
    public function getQty(): float;

    /**
     * Set SKU
     *
     * @param string $sku
     * @return void
     */
    public function setSku(string $sku): void;

    /**
     * Set Quantity
     *
     * @param float $qty
     * @return void
     */
    public function setQty(float $qty): void;

    /**
     * Retrieve existing extension attributes object
     *
     * @return \Magento\InventorySourceSelectionApi\Api\Data\ItemRequestExtensionInterface|null
     */
    public function getExtensionAttributes(): ?ItemRequestExtensionInterface;

    /**
     * Set an extension attributes object
     *
     * @param \Magento\InventorySourceSelectionApi\Api\Data\ItemRequestExtensionInterface $extensionAttributes
     * @return void
     */
    public function setExtensionAttributes(
        \Magento\InventorySourceSelectionApi\Api\Data\ItemRequestExtensionInterface $extensionAttributes
    ): void;
}
```

Each item contains only SKU and Quantity. 
In the default implementation we don't provide any additional data, such as `Shipping address`, because Priority based algorithm no need this data - just the priority of Sources assigned to given Stock matters. 

For more sophisticated algorithms, like a distance-based additional input data may be needed (i.e. Shipping Address), that's why `InventoryRequestInterface` implements ExtensibleDataInterface interfaces and could be extended with these custom input parameters

* Should the products be delivered from the sources designated as having the highest priority? 
* Should the total shipment cost be the primary factor in choosing a shipment method?

> * Should the shipments originate from the closest source?

Reviewer: you could also mention that delivery time could be one of the decisive factors, I know that some merchants interested in the fastest way of delivery as an option

Magento 2.3 provides the priority-based SSA only. Third party developers can create their own algorithms to 

> help merchants decide which shipping option best meets their needs.

Reviewer 2: Might be worth noting they can develop SSAs to run during shipment creation or during checkout. Would also note in the first sentence the priority-based SSA runs during shipment creation.

* Register your SSA within a `di.xml` file

### Implement  `SourceSelectionInterface` 

Your SSA must implement `SourceSelectionInterface`, which is shown below:

```php?start_inline=1
/**
 * Returns source selection algorithm result for given Inventory Request 
 * Current interface should be implemented in order to add own Source Selection Method
 * 
 * @api
 */
interface SourceSelectionInterface
{
    /**
     * @param InventoryRequestInterface $inventoryRequest
     * @return SourceSelectionResultInterface
     */
    public function execute(
       InventoryRequestInterface $inventoryRequest
    ): SourceSelectionResultInterface;
}
````

For example, the following example defines the `MinimalDeliveryCostAlgorithm` class:

```php?start_inline=1
namespace Some\Vendor\Namespace\SourceSelection;

/**
 * Minimal Delivery Cost for Merchant algorithm
 *
 * @api
 */
class MinimalDeliveryCostAlgorithm implements SourceSelectionInterface
{
    public function execute(
       InventoryRequestInterface $inventoryRequest
    ): SourceSelectionResultInterface;
    {
        // TODO: Implement execute() method.
    }
}
```

### Create a `InventoryRequest` factory for quotes  (optional)

> Magento provides the [`InventoryRequestFromOrderFactory`](https://github.com/magento-engcom/msi/blob/2.3.0-release/app/code/Magento/InventoryShipping/Model/InventoryRequestFromOrderFactory.php), which determines which sources to use to full the order at the time a shipment is created. 

Reviewer 2: to use to fulfill the order <-- change full to fulfill

> Magento provides the [`InventoryRequestFromOrderFactory`](https://github.com/magento-engcom/msi/blob/2.3.0-release/app/code/Magento/InventoryShipping/Model/InventoryRequestFromOrderFactory.php), which determines which sources to use to fulfill the order at the time a shipment is created. 

Reviewer 3: which determines which? Is this like where oh where is my grammar book?

Author: Got it. Changed both instances of "which" to "that". 

J/K. I replaced the second "which" to "the"

---
group: inventory
title: Source selection algorithms
---

The **Source Selection Algorithm (SSA)** recommends how to fulfill partial and full shipments. The merchant decides which business needs take precedence when deciding which shipping method to use:

* Should the products be delivered from the sources designated as having the highest priority? 
* Should the total shipment cost be the primary factor in choosing a shipment method?
* Should the shipments originate from the closest source?

> * Should the fastest shipping method be used, even if it's not the cheapest?

Reviewer 2: ```suggestion
* Should the fastest shipping method with the shortest delivery time be used, even if it's not the cheapest?
```

Reviewer 2: Not required, slight addition to support the comment from Igor.

Magento 2.3 provides the priority-based SSA only. Third party developers can create their own algorithms to 
help merchants decide which shipping option best meets their needs.

Magento does not enforce or save the results of SSA recommendations. The recommendations reflect conditions at the moment when the algorithm runs, but conditions change over time. For example, the amount of in-stock products will always fluctuate, and shipping costs might change. The merchant can also modify the recommendations by adjusting quantities for deduction or even by re-assigning the shipment's sources of origin.  

## The priority algorithm

Inventory Management defines the priority algorithm, which recommends delivering products from sources having the highest priority. The `SourceSelectionServiceInterface` accepts an `InventoryRequestInterface` object, which in turn contains the stock ID and a list of items to be shipped. Each item contains only the SKU and quantity. Other potentially relevant data, such as shipping address, is not included, because the priority algorithm does not need it. 

> Additional input data might be needed for more sophisticated algorithms. For example, an algorithm that calculates distance would need the shipping address. that's why `InventoryRequestInterface` implements `ExtensibleDataInterface` interfaces which could be extended with custom input parameters

Reviewer 2: ```suggestion
Additional input data might be needed for more sophisticated algorithms. For example, an algorithm that calculates distance would need the shipping address and all data entered for the source (GPS or full address). That's why `InventoryRequestInterface` implements `ExtensibleDataInterface` interfaces, which can be extended with custom input parameters.
```

Reviewer 2: I mentioned a complete source cause address/GPS is not required. They only need country + zip. The other bits may be needed/useful for a distance algorithm...?

## SSA interfaces

> Currently, Inventory Management deducts stock from the appropriate source after the merchant create a shipment for an order. However, that's not flexible enough--a developer might want to introduce customizations and launch the SSA when the customer proceeds to checkout. Running the SSA at this point could provide the customer more accurate shipping costs. Note in this case, the `Order` object has not created yet, and the system must instead rely on the `Quote` object.

Reviewer 2: ```suggestion
Currently, Inventory Management deducts stock from the appropriate source after the merchant creates a shipment for an order. However, that's not flexible enough--a developer might want to introduce customizations and launch the SSA when the customer proceeds to checkout. Running the SSA at this point could provide the customer more accurate shipping costs. Note in this case, the `Order` object has not created yet, and the system must instead rely on the `Quote` object.
```

* [InventoryRequestInterface](https://github.com/magento-engcom/msi/blob/2.3.0-release/app/code/Magento/InventorySourceSelectionApi/Api/Data/InventoryRequestInterface.php) requests products for a given quantity and stock ID
* [ItemRequestInterface](https://github.com/magento-engcom/msi/blob/2.3.0-release/app/code/Magento/InventorySourceSelectionApi/Api/Data/ItemRequestInterface.php) represents the requested quantity for a specific SKU 
* [SourceSelectionServiceInterface](https://github.com/magento-engcom/msi/blob/2.3.0-release/app/code/Magento/InventorySourceSelectionApi/Api/SourceSelectionServiceInterface.php) returns the source selection algorithm result for the specified `inventoryRequest`
* [GetSourceSelectionAlgorithmListInterface](https://github.com/magento-engcom/msi/blob/2.3.0-release/app/code/Magento/InventorySourceSelectionApi/Api/GetSourceSelectionAlgorithmListInterface.php) returns the list of data interfaces that represent registered SSAs
* [SourceSelectionAlgorithmInterface](https://github.com/magento-engcom/msi/blob/2.3.0-release/app/code/Magento/InventorySourceSelectionApi/Api/Data/SourceSelectionAlgorithmInterface.php) represents a single SSA

> * [SourceSelectionInterface](https://github.com/magento-engcom/msi/blob/2.3.0-release/app/code/Magento/InventorySourceSelectionApi/Model/SourceSelectionInterface.php) returns the SSA result for the specified `inventoryRequest`. 

Reviewer 2: ```suggestion
* [SourceSelectionInterface](https://github.com/magento-engcom/msi/blob/2.3.0-release/app/code/Magento/InventorySourceSelectionApi/Model/SourceSelectionInterface.php) returns the SSA result for the specified `inventoryRequest`
```

Reviewer 2: Removed the period as none of the others have one.

    {
        $requestItems = [];
        $websiteId = $order->getStore()->getWebsiteId();
        $stockId = (int)$this->stockByWebsiteIdResolver->execute((int)$websiteId)->getStockId();
        /** @var OrderItemInterface|OrderItem $orderItem */
        foreach ($order->getItems() as $orderItem) {
            $itemSku = $orderItem->getSku() ?: $this->getSkusByProductIds->execute(
                [$orderItem->getProductId()]
            )[$orderItem->getProductId()];
            $qtyToDeliver = $orderItem->getQtyToShip();
            //check if order item is not delivered yet
            if ($orderItem->isDeleted()
                || $orderItem->getParentItemId()
                || $this->isZero((float)$qtyToDeliver)
                || $orderItem->getIsVirtual()
            ) {
                continue;
            }
            $requestItems[] = $this->itemRequestFactory->create([
                    'sku' => $itemSku,
                    'qty' => $qtyToDeliver
            ]);
        }
        return $this->inventoryRequestFactory->create([
            'stockId' => $stockId,
            'items' => $requestItems
        ]);
    }
}
```

> To add an SSA at the time of checkout, introduce your own factory that produces an  `InventoryRequestInterface` based on a Quote object, outlined as follows:

Reviewer 2: ```suggestion
To add an SSA at the time of checkout, introduce your own factory that produces an  `InventoryRequestInterface` based on a `Quote` object, outlined as follows:
```
